app/services/dropbox/pull_service.rb

```diff
@@ -0,0 +1,111 @@
+module Dropbox
+  class PullService < Operation
+    def execute
+      changes = calculate_changes
+
+      create_or_update(changes[:create_or_update])
+      delete(changes[:delete])
+    end
+
+    private
+
+    def create_or_update(to_create)
+      to_create.each do |props|
+        props[:is_dir] ? create_update_dir(props) : create_update_file(props)
+      end
+    end
+
+    def create_update_dir(props)
+      FileUtils.mkdir_p(props[:local_path])
+      update_entry(props[:path], is_dir: true, revision: props[:revision])
+    end
+
+    def create_update_file(props)
+      download_file(props)
+      update_entry(props[:path],
+                   is_dir: false,
+                   local_hash: calculate_hash(props[:local_path]),
+                   revision: props[:revision])
+    end
+
+    def update_entry(path, attrs = {})
+      app_member.dropbox_entries.
+        find_or_initialize_by(path: path).tap do |entry|
+          entry.assign_attributes(attrs)
+        end.save!
+    end
+
+    def download_file(props)
+      tmp = File.open(new_tmp_path, "wb") do |f|
```

Comment:
Prefer single-quoted strings when you don't need string interpolation or special symbols.